### PR TITLE
Changelog tools

### PR DESCRIPTION
**Pull request by rjmholt**
Adds changelog generation tools to the repo.

The module has some slightly opinionated things about generating a changelog for any repo, and the script generates changelogs for PowerShellEditorServices and vscode-PowerShell and opens PRs for both of them.

All configurable with labelling etc.

### CODE REVIEW

File: tools/GitHubTools.psm1 [hunk 1]
```diff
@@ -1,23 +1,177 @@
...
+    }
+
+    return [GitHubIssueInfo]@{
+        Organization = $IssueUri.Segments[1].TrimEnd('/')
+        Repository = $IssueUri.Segments[2].TrimEnd('/')
+        Number = $issueNum
+    }
+}
+
+filter GetHumanishRepositoryDetails
```
Comment (rkeithhill) [inline tools/GitHubTools.psm1]: LOL!

Comment (rjmholt) [reply tools/GitHubTools.psm1]: `git` seems to like the `ish` suffix to refer to softened parameter "types", like [commitish](https://stackoverflow.com/questions/23303549/what-are-commit-ish-and-tree-ish-in-git) and [humanish](https://stackoverflow.com/questions/13839879/how-to-determine-the-humanish-part-of-a-git-repository)

Comment (rkeithhill) [reply tools/GitHubTools.psm1]: I'm familiar with "commitish" but not "humanish". Thx.

Comment (rjmholt) [reply tools/GitHubTools.psm1]: Aggravatingly they use the terms in passing and you have to sift through StackOverflow to decode them

File: tools/ChangelogTools.psm1 [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+                $changelogItem.IssueNumber = $closedIssueInfos[0].Number
+                $changelogItem.Labels += ($closedIssueInfos | ForEach-Object { $_.Labels })
+            }
+        }
+
+        $changelogItem
+    }
+}
+
+Export-ModuleMember -Function Get-ChangelogItemFromCommit
```
Comment (TylerLeonhardt) [inline tools/ChangelogTools.psm1]:
```suggestion
Export-ModuleMember -Function Get-ChangelogItemFromCommit

```

**Comment by TylerLeonhardt** [comment]
Oh?

**Comment by rjmholt** [comment]
Ah that's because I accidentally deleted the branch!

**Comment by rjmholt** [comment]
Ok here's the kinds of PR this opens:

- [PSES](https://github.com/rjmholt/PowerShellEditorServices/pull/10/files)
- [vscode-powershell](https://github.com/rjmholt/vscode-powershell/pull/15/files)

A few notes:

- This adds a little emoji based on the issue label, if there is one
- The link is always to the PR, but if it closes an issue the issue is used as the name. When vscode-powershell issues are closed in PSES, it says `[vscode-PowerShell #XXXX]`
- If there's no PR for a commit, there's no `[...]` at the start
- This depends on the GitHub commit message format where it adds `(#XXXX)` to the end of commit messages
- We can easily add more tags in various places:
  - PR labels like PowerShell's <kbd>CL-Breaking</kbd> tags
  - Issue labels, like the existing <kbd>Issue-Bug</kbd>, <kbd>Issue-Enhancement</kbd>
  - In commits with `[tag]` -- currently `[Ignore]` is registered

The changelog script also separates the changelog into sections if any PRs close categorised issues (e.g. <kbd>Area-PSReadLine<kbd>). Again we can easily add to this with PR labels too.

File: tools/changelog/updateChangelog.ps1 [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    DefaultCategory = $defaultCategory
+    ReleaseName = $ReleaseName
+    DateFormat = $dateFormat
+}
+
+$psesChangelogSection = Get-GitCommit -SinceRef $SinceRef -UntilRef $UntilRef -GitHubToken $GitHubToken -RepositoryPath $PsesRepositoryPath |
+    Get-ChangeInfoFromCommit -GitHubToken $GitHubToken |
+    Skip-IgnoredChange @ignore |
+    New-ChangelogEntry @clEntryParams |
+    New-ChangelogSection @clSectionParams -Preamble "#### [$psesRepoName](https://github.com/$Organization/$psesRepoName)"
```
Comment (TylerLeonhardt) [inline tools/changelog/updateChangelog.ps1]: PSES does not have this Preamble

File: tools/ChangelogTools.psm1 [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+                $changelogItem.ClosedIssues = $closedIssueInfos | Get-GitHubIssue
+                $changelogItem.IssueNumber = $closedIssueInfos[0].Number
+            }
+        }
+
+        $changelogItem
+    }
+}
+
+filter New-ChangelogEntry
```
Comment (corbob) [inline tools/ChangelogTools.psm1]: I don't know how important making Codacy green is here, but could add something like: `[Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSUseShouldProcessForStateChangingFunctions", "", Justification="We're not actually changing system state.")]` to suppress it complaining about failing the PSSA rule...

Comment (rjmholt) [reply tools/ChangelogTools.psm1]: Ironically there are much more dangerous methods in the GitHubTools.psm1 module that it misses. PSScriptAnalyzer has some weird bits in it frankly -- I'd love to make it more logical.

File: tools/GitHubTools.psm1 [hunk 5]
```diff
@@ -282,7 +285,14 @@ function Copy-GitRepository
         New-Item -Path $containingDir -ItemType Directory -ErrorAction Stop
     }
 
-    Exec { git clone --single-branch --branch $CloneBranch $OriginRemote $Destination }
+    Write-Verbose "Cloning git repository '$OriginRemote' to path '$Destination'"
+
+    if ($CloneBranch)
+    {
+        Write-Verbose "Cloned branch: $CloneBranch"
```
Comment (TylerLeonhardt) [inline tools/GitHubTools.psm1]: shouldn't this be after the exec?

File: tools/changelog/updateChangelog.ps1 [hunk 6]
```diff
@@ -6,22 +6,55 @@
 using module ..\GitHubTools.psm1
 using module ..\ChangelogTools.psm1
 
+<#
+.EXAMPLE
+.\updateChangelog.ps1 -GitHubToken $ghTok -PSExtensionSinceRef v2.0.0-preview.3 -PsesSinceRef v2.0.0-preview.3 -PSExtensionVersion 2019.5.0 -PsesVersion 1.12.2 -PSExtensionUntilRef legacy/1.x -PsesUntilRef legacy/1.x -Verbose
```
Comment (TylerLeonhardt) [inline tools/changelog/updateChangelog.ps1]: I'm confused by this example. You have a mix of preview and legacy?

File: tools/changelog/updateChangelog.ps1 [hunk 7]
```diff
@@ -111,101 +201,150 @@ $categories = [ordered]@{
...
+    UntilRef = $PsesUntilRef
+    GitHubToken = $GitHubToken
+    RepositoryPath = $PsesRepositoryPath
+    Verbose = $VerbosePreference
+}
+
 $clEntryParams = @{
     EntryCategories = $categories
     DefaultCategory = $defaultCategory
     TagLabels = @{
```
Comment (TylerLeonhardt) [inline tools/changelog/updateChangelog.ps1]: 💙

File: tools/changelog/updateChangelog.ps1 [hunk 8]
```diff
@@ -111,101 +201,150 @@ $categories = [ordered]@{
...
 
-#endregion
+#endregion PSES Changelog
 
 #region vscode-PowerShell Changelog
-
-$psesChangelogPostamble = ($psesChangelogSection -split "`n")
-$psesChangelogPostamble = $psesChangelogPostamble[2..$psesChangelogPostamble.Length]
+$psesChangelogPostamble = $psesChangelogSection -split "`n"
+$psesChangelogPostamble = @("#### [$psesRepoName](https://github.com/$Organization/$psesRepoName)") + $psesChangelogPostamble[2..($psesChangelogPostamble.Length-3)]
```
Comment (TylerLeonhardt) [inline tools/changelog/updateChangelog.ps1]: It'd be nice if you added a small comment about these numbers ( 2 and 3)

**Pull request merged**